### ISSUE TITLE
Consider emmet completions from php extension

### ISSUE THREAD

**Issue by ramya-rao-a** [Microsoft/vscode#29354]
See #29318 and https://github.com/Microsoft/vscode/issues/28286#issuecomment-310749096

**Comment by jens1o** [comment]
Something new, would like to see it in action! :)

**Comment by ramya-rao-a** [comment]
@roblourens and I discussed this and it looks tough at the moment.
The helper is node module and php language service I believe is in php. 
How do we make them talk to each other?

**Comment by jens1o** [comment]
What about websockets? For php I use this library: http://socketo.me/

**Comment by ramya-rao-a** [comment]
I am not familiar with that, but you are welcome to give it a try :)

**Comment by roblourens** [comment]
Spawning a node process from PHP is probably one solution. Making them talk to each other is easy, websockets not needed.

**Comment by ramya-rao-a** [comment]
cc @felixfbecker for this thoughts

**Comment by felixfbecker** [comment]
I think this is the same issue as HTML suggestions inside PHP in general (or CSS in HTML, or JS in HTML, or SQL in PHP). Imo requiring the language server to handle all possible embedded languages (and the possibly embedded languages inside the embedded language and so on) doesn't scale, because for N language servers you will each have to build and integrate M<N embedded languages, which can then have embedded languages again, so in the worst case you have to embed close to N^N language servers. For example, PHP needs to embed HTML, but HTML then needs to embed CSS and JS. The user also can't decide anymore what language server to use for the embedded languages. Overall, we would end up with a very inconsistent experience across languages.

A better approach I think is what I outlined in this thread: https://github.com/Microsoft/vscode/issues/670#issuecomment-282344532
In summary, let the language server declare ranges of embedded languages to the client and then let the client (VS Code) use the appropriate language server for that range+language.

**Comment by bmewburn** [comment]
A little off topic but how do I turn emmet off? It's popping up everywhere and often the first selection in the list.
![screenshot from 2017-08-16 12-11-03](https://user-images.githubusercontent.com/5885195/29344586-2f3148d8-827c-11e7-8b82-e3e7f11a8538.png)

**Comment by ramya-rao-a** [comment]
@bmewburn set `emmet.showExpandedAbbreviation` to `inMarkupAndStylesheetFilesOnly` or `never`

However, we have refined the triggering of emmet suggestions in the latest Insiders (https://code.visualstudio.com/insiders). The same will be out in tomorrow's release of 1.15.1

So if you use emmet, I would suggest to set `emmet.showExpandedAbbreviation` back to `always` once 1.15.1 is released

**Comment by mcdado** [comment]
@ramya-rao-a I'm using 1.18.1 but I still need to set `emmet.showExpandedAbbreviation` to `inMarkupAndStylesheetFilesOnly` to solve this problem. Am I doing something wrong?

**Comment by ramya-rao-a** [comment]
@mcdado To solve which problem? Please share screenshots

**Comment by mcdado** [comment]
I admit that when I was figuring it out I thought it was happening exactly like [here](https://github.com/Microsoft/vscode/issues/29354#issuecomment-322644442), with random keywords, but I now realize it is different, it happens with HTML keywords:

![screen shot 2017-12-13 at 20 59 22](https://user-images.githubusercontent.com/898057/33959718-aaebcca8-e048-11e7-80df-873c0b3d2b56.png)

This is without `inMarkupAndStylesheetFilesOnly`.

**Comment by bmewburn** [comment]
With the [intelephense](https://marketplace.visualstudio.com/items?itemName=bmewburn.vscode-intelephense-client) extension my preference is to set `emmet.showExpandedAbbreviation` to `inMarkupAndStylesheetFilesOnly` as the extension will forward requests on to other extensions listening to `html` when outside of `<?php ?>`, meaning you will get emmet suggestions in a more appropriate context.

**Comment by ramya-rao-a** [comment]
@mcdado Can't do much for cases like that other than setting `emmet.showExpandedAbbreviation` to `inMarkupAndStylesheetFilesOnly` and relying on using tab for emmet instead by setting 
 `emmet.triggerExpansionOnTab` to `true`

@bmewburn I did not know that, good to know! Can you point me to the code that does that?

**Comment by bmewburn** [comment]
@ramya-rao-a here you go https://github.com/bmewburn/vscode-intelephense/blob/master/client/src/embeddedContentDocuments.ts 😄

**Comment by mcdado** [comment]
@bmewburn Indeed I'm using your awesome extension. I added again `inMarkupAndStylesheetFilesOnly` and it works great, indeed.
